exec/bigmachine_test.go

```diff
@@ -171,6 +171,129 @@ func TestBigmachineExecutorPanicCompile(t *testing.T) {
 	run(t, x, tasks, TaskErr)
 }
 
+// TestBigmachineExecutorProcs verifies that using the Procs pragma properly
+// affects machine/proc allocation.
+func TestBigmachineExecutorProcs(t *testing.T) {
+	// Set up the test with:
+	// - a slice with 8 tasks
+	// - Procs(2) so that two procs are allocated for each task
+	// - a system with 12 procs, 4 procs per machine
+	system := testsystem.New()
+	system.Machineprocs = 4
+	ctx, cancel := context.WithCancel(context.Background())
+	x := newBigmachineExecutor(system)
+	shutdown := x.Start(&Session{
+		Context: ctx,
+		p:       12, // 3 machines
+		maxLoad: 1,
+	})
+	defer shutdown()
+	defer cancel()
+
+	// We use the wait group to block completion of tasks, controlling execution
+	// for our test.
+	var wg sync.WaitGroup
+	fn := bigslice.Func(func() bigslice.Slice {
+		is := make([]int, 100)
+		for i := range is {
+			is[i] = i
+		}
+		slice := bigslice.ReaderFunc(8, func(shard int, x *int, xs []int) (int, error) {
+			wg.Wait()
+			const N = 10
+			var i int
+			for *x < N && i < len(xs) {
+				xs[i] = (shard * N) + *x
+				i++
+				*x++
+			}
+			if *x == N {
+				return i, sliceio.EOF
+			}
+			return i, nil
+		})
+		// Add an identity mapping to exercise pipelining.
+		slice = bigslice.Map(slice, func(i int) int {
+			return i
+		}, bigslice.Procs(2))
+		return slice
+	})
+	inv := fn.Invocation("<test>")
+	slice := inv.Invoke()
+	tasks, err := compile(slice, inv, false)
+	if err != nil {
+		t.Fatal(err)
+	}
+	// Verify that there is one task per shard.
+	if got, want := len(tasks), 8; got != want {
+		t.Fatalf("got %v, want %v", got, want)
+	}
+	// Verify that the proc need is propagated through the pipeline.
+	for _, task := range tasks {
+		if got, want := task.Pragma.Procs(), 2; got != want {
+			t.Fatalf("got %v, want %v", got, want)
+		}
+	}
+	// Don't let anything run to completion yet.
+	wg.Add(1)
+	// Run three tasks (needing 6 procs), and verify that two machines have been
+	// started on which to run them.
+	for _, task := range tasks[:3] {
+		go x.Run(task)
+		state, err := task.WaitState(ctx, TaskRunning)
+		if err != nil || state != TaskRunning {
+			t.Fatal(state, err)
+		}
+	}
+	if got, want := system.N(), 2; got != want {
+		t.Errorf("got %v, want %v", got, want)
+	}
+	// Run the rest of the tasks, and verify that the remaining machines have
+	// been started on which to run them.
+	//
+	// Note: this is racy, as we don't have a way of knowing that the executor
+	// has blocked because it cannot acquire a machine on which to run a task.
+	// If this is a problem, we'll need a better solution.
+	for _, task := range tasks[3:] {
+		go x.Run(task)
+		func() {
+			ctx, cancel := context.WithTimeout(ctx, 5*time.Millisecond)
```

Comment:
Optional: This seems short enough that it might be lost in scheduling overhead on a busy CI machine. If I understand correctly, this is a tradeoff between detecting slow task start to avoid misleading test failures vs. making the test slower overall. If that's correct, maybe 100 ms is reasonable?